internal/gatewayapi/route.go

```diff
@@ -438,6 +427,46 @@ func (t *Translator) processGRPCRouteRule(grpcRoute *GRPCRouteContext, ruleIdx i
 	return ruleRoutes
 }
 
+func (t *Translator) processGRPCRouteMethodExact(method *v1alpha2.GRPCMethodMatch, irRoute *ir.HTTPRoute) {
+	switch {
+	case method.Service != nil && method.Method != nil:
+		irRoute.PathMatch = &ir.StringMatch{
+			Exact: StringPtr(fmt.Sprintf("/%s/%s", *method.Service, *method.Method)),
+		}
+	case method.Method != nil:
+		// Use a header match since the PathMatch doesn't support Suffix matching
+		irRoute.HeaderMatches = append(irRoute.HeaderMatches, &ir.StringMatch{
+			Name:   ":path",
+			Suffix: StringPtr(fmt.Sprintf("/%s", *method.Method)),
+		})
+	case method.Service != nil:
+		irRoute.PathMatch = &ir.StringMatch{
+			Prefix: StringPtr(fmt.Sprintf("/%s", *method.Service)),
+		}
+	}
+}
+
+func (t *Translator) processGRPCRouteMethodRegularExpression(method *v1alpha2.GRPCMethodMatch, irRoute *ir.HTTPRoute) {
+	switch {
+	case method.Service != nil && method.Method != nil:
+		irRoute.PathMatch = &ir.StringMatch{
+			SafeRegex: StringPtr(fmt.Sprintf("/%s/%s", *method.Service, *method.Method)),
+		}
+	case method.Method != nil:
+		irRoute.PathMatch = &ir.StringMatch{
+			SafeRegex: StringPtr(fmt.Sprintf("/.+/%s", *method.Method)),
```

Comment:
thinking out loud, it is okay to use `.+` here (accepts one or more of any char) because the field has already been validated in the validation webhook, before it reaches here ?

Reply:
actually this field never be validated by validation webhook when `MatchType=RegularExpression`, see https://github.com/kubernetes-sigs/gateway-api/blob/8dde50805b8010f6ec52ae52e7d47d799cdcaabf/apis/v1alpha2/validation/grpcroute.go#L80

so we need `/.+/%s` to make sure that the `service` has been matched at right position, for example, not `//method` nor `/method`, but `/x/method`.

and so does `/%s/.+`, making sure the `method` has been matched at right position.

Reply:
thanks for checking upstream, in that case should this be `[A-Za-z0-9]+` instead of `.+` so we are validating char type

Reply:
sure that should work for both, thanks !
can you please test this out as well, here is a good starting point https://gateway.envoyproxy.io/latest/user/grpc-routing.html

Reply:
hi, I have run some test on `/[validServiceName]/%s` and `/%s/[validMethodName]` case, and it works well.

here is a snapshot about `/[validServiceName]/%s` case:

![Snipaste_2023-04-08_10-43-58](https://user-images.githubusercontent.com/40317966/230700169-47a4887d-7348-4f6d-b03f-7d3ab8e8fec1.jpg)

Reply:
can you try by using a regex value for service such as `*.Echo` and leaving `method` to be `nil` ?

Reply:
cannot do this now, present gateway-api v0.6.2 will determine this as an error: https://github.com/kubernetes-sigs/gateway-api/blob/7042cd866bc69b160ab6d04b51676e14b487505d/apis/v1alpha2/grpcroute_types.go#L329

the changes in https://github.com/kubernetes-sigs/gateway-api/issues/1746 will be introduced in gateway-api v0.7.0

Reply:
can be tested until this https://github.com/envoyproxy/gateway/issues/1267 solved